### PR TITLE
[Feature] Support RMosaic

### PR DESCRIPTION
**Pull request by heiyuxiaokai**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Please describe the motivation of this PR and the goal you want to achieve through this PR.

## Modification
1.Support RMosaic
(PS: Corresponding codes in mmdetection shows Mosaic must be used with `RandomAffine`. But I find `Resize` is also working. )
## BC-breaking (Optional)

Does the modification introduce changes that break the back-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)
Use the train_pipeline as belowing:
`train_pipeline = [
    dict(type='RMosaic', img_scale=(800, 800)),
    dict(type='RResize', img_scale=(800, 800)),
    dict(
        type='RRandomFlip',
        flip_ratio=[0.25, 0.25, 0.25],
        direction=['horizontal', 'vertical', 'diagonal'],
        version=angle_version),
    dict(
        type='PolyRandomRotate',
        rotate_ratio=0.5,
        angles_range=180,
        auto_bound=False,
        # rect_classes=[9, 11],
        version=angle_version),
    dict(type='Normalize', **img_norm_cfg),
    dict(type='Pad', size_divisor=32),
    dict(type='DefaultFormatBundle'),
    dict(type='Collect', keys=['img', 'gt_bboxes', 'gt_labels'])
]`

`train_dataset = dict(
    _delete_ = True,
    type='MultiImageMixDataset',
    dataset=dict(
        type=dataset_type,
        classes=classes,
        ann_file=data_root + 'trainval/annfiles/',
        img_prefix=data_root + 'trainval/images/',
        pipeline=[
            dict(type='LoadImageFromFile'),
            dict(type='LoadAnnotations', with_bbox=True)
        ],
        version=angle_version
    ),
    pipeline=train_pipeline
)`

`data = dict(
    samples_per_gpu=4,
    workers_per_gpu=2,
    train=train_dataset,
    val=dict(
        classes=classes,
        ann_file=data_root + 'test/annfiles/',
        img_prefix=data_root + 'test/images/',
        pipeline=test_pipeline,
        version=angle_version),
    test=dict(
        classes=classes,
        ann_file=data_root + 'test/annfiles/',
        img_prefix=data_root + 'test/images/',
        pipeline=test_pipeline,
        version=angle_version))`

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

**Comment by zytx121** [comment]
Thank you for your PR! It's cool!
Could you please add the docstring for these functions?

**Comment by heiyuxiaokai** [comment]
> Thank you for your PR! It's cool! Could you please add the docstring for these functions?



> Thank you for your PR! It's cool! Could you please add the docstring for these functions?

Ok, I'll update it later.

**Comment by liuyanyi** [comment]
Please follow https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md to set pre-commit hook and fix the lint error.

**Comment by heiyuxiaokai** [comment]
> Please follow https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md to set pre-commit hook and fix the lint error.

Thanks! I'm confused about that.

File: mmrotate/datasets/pipelines/transforms.py [hunk 1]
```diff
@@ -361,3 +365,180 @@ def _crop_data(self, results, crop_size, allow_negative_crop):
                 results[label_key] = results[label_key][valid_inds]
 
         return results
+
+
+@ROTATED_PIPELINES.register_module()
+class RMosaic(Mosaic):
+    """Rotate Mosaic augmentation. Inherit from
+    'mmdet.datasets.pipelines.transforms.Mosaic'.
```
Comment (zytx121) [inline mmrotate/datasets/pipelines/transforms.py:373]:
```suggestion
    `mmdet.datasets.pipelines.transforms.Mosaic`.
```

File: mmrotate/datasets/pipelines/transforms.py [hunk 2]
```diff
@@ -361,3 +365,180 @@ def _crop_data(self, results, crop_size, allow_negative_crop):
...
+     The mosaic transform steps are as follows:
+         1. Choose the mosaic center as the intersections of 4 images
+         2. Get the left top image according to the index, and randomly
+            sample another 3 images from the custom dataset.
+         3. Sub image will be cropped if image is larger than mosaic patch
+
+    Args:
+        img_scale (Sequence[int]): Image size after mosaic pipeline of single
+            image. The shape order should be (height, width).
+            Default to (640, 640).
```
Comment (zytx121) [inline mmrotate/datasets/pipelines/transforms.py:406]:
```suggestion
            Defaults to (640, 640).
```

File: mmrotate/datasets/pipelines/transforms.py [hunk 3]
```diff
@@ -361,3 +365,180 @@ def _crop_data(self, results, crop_size, allow_negative_crop):
...
+        min_bbox_size (int | float): The minimum pixel for filtering
+            invalid bboxes after the mosaic pipeline. Default to 0.
+        bbox_clip_border (bool, optional): Whether to clip the objects outside
+            the border of the image. In some dataset like MOT17, the gt bboxes
+            are allowed to cross the border of images. Therefore, we don't
+            need to clip the gt bboxes in these cases. Defaults to True.
+        skip_filter (bool): Whether to skip filtering rules. If it
+            is True, the filter rule will not be applied, and the
+            `min_bbox_size` is invalid. Default to True.
+        pad_val (int): Pad value. Default to 114.
```
Comment (zytx121) [inline mmrotate/datasets/pipelines/transforms.py:418]:
```suggestion
        pad_val (int): Pad value. Defaults to 114.
```

File: mmrotate/datasets/pipelines/transforms.py [hunk 4]
```diff
@@ -361,3 +365,180 @@ def _crop_data(self, results, crop_size, allow_negative_crop):
...
+         3. Sub image will be cropped if image is larger than mosaic patch
+
+    Args:
+        img_scale (Sequence[int]): Image size after mosaic pipeline of single
+            image. The shape order should be (height, width).
+            Default to (640, 640).
+        center_ratio_range (Sequence[float]): Center ratio range of mosaic
+            output. Default to (0.5, 1.5).
+        min_bbox_size (int | float): The minimum pixel for filtering
+            invalid bboxes after the mosaic pipeline. Default to 0.
```
Comment (zytx121) [inline mmrotate/datasets/pipelines/transforms.py:410]:
```suggestion
            invalid bboxes after the mosaic pipeline. Defaults to 0.
```

File: mmrotate/datasets/pipelines/transforms.py [hunk 5]
```diff
@@ -361,3 +365,180 @@ def _crop_data(self, results, crop_size, allow_negative_crop):
...
+         2. Get the left top image according to the index, and randomly
+            sample another 3 images from the custom dataset.
+         3. Sub image will be cropped if image is larger than mosaic patch
+
+    Args:
+        img_scale (Sequence[int]): Image size after mosaic pipeline of single
+            image. The shape order should be (height, width).
+            Default to (640, 640).
+        center_ratio_range (Sequence[float]): Center ratio range of mosaic
+            output. Default to (0.5, 1.5).
```
Comment (zytx121) [inline mmrotate/datasets/pipelines/transforms.py:408]:
```suggestion
            output. Defaults to (0.5, 1.5).
```

File: mmrotate/datasets/pipelines/transforms.py (hunk 4 above)
Comment (zytx121) [inline mmrotate/datasets/pipelines/transforms.py:410]:
```suggestion
            invalid bboxes after the mosaic pipeline. Defaults to 0.
```

File: mmrotate/datasets/pipelines/transforms.py [hunk 6]
```diff
@@ -361,3 +365,180 @@ def _crop_data(self, results, crop_size, allow_negative_crop):
...
+        min_bbox_size (int | float): The minimum pixel for filtering
+            invalid bboxes after the mosaic pipeline. Default to 0.
+        bbox_clip_border (bool, optional): Whether to clip the objects outside
+            the border of the image. In some dataset like MOT17, the gt bboxes
+            are allowed to cross the border of images. Therefore, we don't
+            need to clip the gt bboxes in these cases. Defaults to True.
+        skip_filter (bool): Whether to skip filtering rules. If it
+            is True, the filter rule will not be applied, and the
+            `min_bbox_size` is invalid. Default to True.
```
Comment (zytx121) [inline mmrotate/datasets/pipelines/transforms.py:417]:
```suggestion
            `min_bbox_size` is invalid. Defaults to True.
```

File: mmrotate/datasets/pipelines/transforms.py [hunk 7]
```diff
@@ -361,3 +365,180 @@ def _crop_data(self, results, crop_size, allow_negative_crop):
...
+        pad_val (int): Pad value. Default to 114.
+        prob (float): Probability of applying this transformation.
+            Default to 1.0.
+    """
```
Comment (zytx121) [inline mmrotate/datasets/pipelines/transforms.py:421]: Version's docstring is missing.

**Comment by heiyuxiaokai** [comment]
Mosaic-transform outputs twice the input images. And resize-transform convert them to original sizes, which means all gt-boxes are half of inputs'. So I choose to set the min_box_size to 10, corresponding to min_box_size of randomrotate. 
From the view of visualizations, there seems no problem. But I got bad performance as：
image

Maybe mosaic should be used with randomcrop, and I'll try it.

**Comment by heiyuxiaokai** [comment]
Thanks for your explaining @liuyanyi  . I'll have a try.

**Review by zytx121 — APPROVED** [review]
LGTM

**Comment by nanhui69** [comment]
> Thanks for your explaining @liuyanyi . I'll have a try.

does mosaic transform should combine with mixup? if that ,could you give a sample?
